Add runbook for changing storage node size on Exoscale (#101)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/exoscale/change_storage_node_size.adoc

* `commodore`, see https://syn.tools/commodore/running-commodore.html[Running Commodore]
* `jq`
* `yq` https://mikefarah.gitbook.io/yq[yq YAML processor] (version 4 or higher)

== Prepare local environment

. Create local directory to work in
+
[TIP]
====
We strongly recommend creating an empty directory, unless you already have a work directory for the cluster you're about to work on.
This guide will run Commodore in the directory created in this step.
====
+
[source,bash]
----
export WORK_DIR=/path/to/work/dir
mkdir -p "${WORK_DIR}"
pushd "${WORK_DIR}"
----

. Configure API access
+
include::partial$exoscale/environment-vars.adoc[]
+
include::partial$vshn-input.adoc[]

. Get required tokens from Vault
+
include::partial$connect-to-vault.adoc[]
+
include::partial$get-hieradata-token-from-vault.adoc[]

. Compile the catalog for the cluster.
Having the catalog available locally enables us to run Terraform for the cluster to make any required changes.
+
[source,bash]
----
commodore catalog compile "${CLUSTER_ID}"
----

== Prepare Terraform environment

include::partial$exoscale/configure-terraform-secrets.adoc[]

include::partial$setup_terraform.adoc[]

> 

Author: My thinking was to ensure working terraform before changing anything. No strong opinions.

----

. Disable auto sync for component `rook-ceph`.
This allows us to temporarily make manual changes to the Rook Ceph cluster.
+
include::partial$disable-argocd-autosync.adoc[]

. Make a note of the original count of OSDs in the Ceph cluster
+
[source,bash]
----
orig_osd_count=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster \
  get cephcluster cluster -o jsonpath='{.spec.storage.storageClassDeviceSets[0].count}')
echo $orig_osd_count
----

. Change Ceph cluster to lauch new OSDs
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster patch cephcluster cluster --type=json \
  -p "[{
    \"op\": \"replace\",
    \"path\": \"/spec/storage/storageClassDeviceSets/0/count\",
    \"value\": $(expr ${orig_osd_count} '*' 2)
  }]"
----

. Wait until the new OSDs are launched
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster get pods -w

# Wait for all PGs to reach `active+clean` state
watch kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
  ceph -s
----

=== Remove old OSD

. Tell Ceph to take OSDs on old nodes out of service and relocate data stored on them
+

> [source,bash]
> ----
> for node in $(echo -n $NODES_TO_REPLACE); do
>   osd_id=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
>     -l failure-domain="${node}" --no-headers \
>     -o custom-columns="NAME:.metadata.name" | cut -d- -f4)
>   echo Taking OSD out of service:
>   kubectl --as=cluster-admin -n syn-rook-ceph-cluster get pods -o wide | grep "osd-${osd_id}"
>   echo -n "Is the node correct? (y/N) "; read REPLY; echo
>   if [[ $REPLY =~ ^[Yy]$ ]]; then
>       kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
>         ceph osd crush reweight "osd.${osd_id}" 0
>   fi
> done
> ----

Reviewer: This seems a bit overengineered for a non-destructive operation, I would probably have just left it at

```suggestion
[source,bash]
----
# Verify that the list of nodes to replace is correct
echo $NODES_TO_REPLACE
# Reweight OSDs on those nodes to 0
for node in $(echo -n $NODES_TO_REPLACE); do
  osd_id=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
    -l failure-domain="${node}" --no-headers \
    -o custom-columns="OSD_ID:.metadata.labels.ceph_daemon_id")
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
    ceph osd crush reweight "osd.${osd_id}" 0
done
----
```

. Wait for the data to be redistributed ("backfilled") to the other OSDs.
+
include::partial$storage-ceph-backfilling.adoc[]

> . Remove the OSD from the Ceph cluster

Reviewer 2: At this point the OSDs are not set to out. I guess this is fine? But maybe this time we actually need the `--yes-i-really-mean-it`?

Author: We take the nodes out with "ceph osd crush reweight" which is recommended ["sometimes"](https://docs.ceph.com/en/latest/rados/operations/add-or-rm-osds/) for smaller clusters.

Reviewer 2: Hmm oke, then it's fine. I did not expect the `crush reweight` command to actually mark the OSD as `out`.

+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do

>   osd_id=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
>     -l failure-domain="${node}" --no-headers \
>     -o custom-columns="NAME:.metadata.name" | cut -d- -f4)

Reviewer: ```suggestion
  osd_id=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
    -l failure-domain="${node}" --no-headers \
    -o custom-columns="OSD_ID:.metadata.labels.ceph_daemon_id")
```

  kubectl --as=cluster-admin -n syn-rook-ceph-cluster scale --replicas=0 \
    "deploy/rook-ceph-osd-${osd_id}"
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \

>     ceph osd purge "${osd_id}" --yes-i-really-mean-it

Reviewer: Is `--yes-i-really-mean-it` actually required here?

Author: I'll test it in the final round of testing 👍

  kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
    ceph osd crush remove "${node}"
done
----

. Check that the OSDs are no longer listed in `ceph osd tree`
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
  ceph osd tree
----

. Scale down the Rook-Ceph operator
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-operator scale --replicas=0 \
  deploy/rook-ceph-operator
----

> . Make a note of the PVC of the old OSD

Reviewer 2: A little pedantic, but that should be in plural. There are a few other places that are not quite consistent.

Author: went through the document and tried to fix all places where I missed it...

+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete deploy \
    -l failure-domain="${node}"
done
----

. Reset Ceph cluster resource to have original number of OSDs
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster patch cephcluster cluster --type=json \
  -p "[{
    \"op\": \"replace\",
    \"path\": \"/spec/storage/storageClassDeviceSets/0/count\",
    \"value\": ${orig_osd_count}
  }]"
----

. Clean up PVCs and prepare jobs of the old OSDs if necessary
+
[source,bash]
----
for pvc_name in $(echo -n $old_pvc_names); do
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete job \
    -l ceph.rook.io/pvc="${pvc_name}"
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete pvc "${pvc_name}"
done
----

. Clean up PVC encryption secrets
+
[source,bash]
----
for pvc_name in $(echo -n $old_pvc_names); do
  kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete secret -l pvc_name="${pvc_name}"
done
----

. Scale up the Rook-Ceph operator
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-operator scale --replicas=1 \
  deploy/rook-ceph-operator
----

> === Remove the old MONs

Reviewer: ```suggestion
=== Move the MONs to the new nodes
```

> . Find the MONs (if any) on the nodes to be replaced

Reviewer: There will always be MONs that need to be moved for this how-to.

```suggestion
. Find the MONs on the nodes to be replaced
```

+
[source,bash]
----
MON_IDS=""
for node in $(echo -n $NODES_TO_REPLACE); do
  mon_id=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get pods \
    -lapp=rook-ceph-mon -o wide \
    | grep "${node}" | cut -d- -f4)
  if [[ -n "$mon_id" ]]; then
    MON_IDS="$MON_IDS $mon_id"

>   fi

Reviewer 2: I don't think that is necessary? This list is only read by humans so an extra space shouldn't matter.

Author: You're right. I tried automating it first but it's a tad risky. I removed them all at once.

----
+

> TIP: You can skip the remaining steps in this section if `$MON_IDS` is empty.

Reviewer: This can be dropped, if anything, there should be a note saying that there **must** be MONs that need to be moved for this procedure.

====
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster patch cephcluster cluster --type=json \
  -p '[{
    "op": "replace",
    "path": "/spec/healthCheck/daemonHealth/mon",
    "value": {
      "disabled": false,
      "interval": "10s",
      "timeout": "10s"
    }
  }]'
kubectl --as=cluster-admin -n syn-rook-ceph-operator delete pods \
  -l app=rook-ceph-operator
----

. Wait for operator to settle.
Wait for a log message saying `done reconciling ceph cluster in namespace "syn-rook-ceph-cluster"`
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-operator logs -f \
  deploy/rook-ceph-operator
----

. Cordon old nodes
+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do
  kubectl --as=cluster-admin cordon "${node}"
done
----
+
. For every id in `$MON_IDS` replace the MON pod
+
[source,bash]
----
mon_id=<MON_ID>
kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete pod \
  -l app=rook-ceph-mon,ceph_daemon_id="${mon_id}"

# Wait until new MON is scheduled
kubectl --as=cluster-admin -n syn-rook-ceph-cluster get pods -w

> kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
>   ceph status

Reviewer: ```suggestion
# Wait until the cluster has regained full quorum
kubectl --as=cluster-admin -n syn-rook-ceph-cluster exec -it deploy/rook-ceph-tools -- \
  ceph status
```

Once https://github.com/rook/rook/issues/8363[Rook.io GitHub issue #8363] is fixed, the operator restart shouldn't be necessary anymore.
====
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster patch cephcluster cluster --type=json \
  -p '[{
    "op": "replace",
    "path": "/spec/healthCheck/daemonHealth/mon",
    "value": {}
  }]'
kubectl --as=cluster-admin -n syn-rook-ceph-operator delete pods \
  -l app=rook-ceph-operator
----

=== Clean up the old nodes

. Drain the nodes to replace
+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do
  kubectl --as=cluster-admin drain "${node}" \
    --delete-emptydir-data --ignore-daemonsets
done
----

. Delete the nodes from the cluster
+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do
  kubectl --as=cluster-admin delete node "${node}"
done
----

. Remove the Exoscale VMs
+
[source,bash]
----
for node in $(echo -n $NODES_TO_REPLACE); do
  node_id=$(exo vm list -O json | \
    jq --arg storage_node "$node" -r \
    '.[] | select(.name==$storage_node) | .id')

  echo "Removing node:"
  exo vm list | grep "${node_id}"
  echo -n "Is this correct? (y/N) "; read REPLY; echo

>   if [[ $REPLY =~ ^[Yy]$ ]]; then

Reviewer: Unnecessary, the `exo` CLI itself asks whether to remove the node.

  echo -n "Is this correct? (y/N) "; read REPLY; echo
  if [[ $REPLY =~ ^[Yy]$ ]]; then
    exo vm delete "${node_id}"
  fi
done
----

. Clean up localstorage PV of decommissioned node
+
[source,bash]
----
kubectl --as=cluster-admin delete pv "${pv_name}"

> ----

Reviewer 2: There is more than just one PV. You probably meant to iterate over `old_pv_names`
